ext/drreg/drreg.c

```diff
@@ -2042,203 +2008,181 @@ static bool
+     */
+    for (inst = instrlist_last(info->fragment_info.ilist);
+         inst != NULL && instr_get_next(inst) != fault_inst;
+         inst = instr_get_prev(inst)) {
+        /* XXX: The check for faux spills/restores (app instrs that look like drreg
+         * spills/restores) needs the ilist, but maybe we can live without that check,
+         * or atleast merge drreg_event_restore_state_with_ilist and
+         * drreg_event_restore_state_without_ilist and check for !instr_is_app only if
+         * the ilist is available.
          */
         if (!instr_is_app(inst)) {
-            if (is_our_spill_or_restore(drcontext, inst, &spill, &reg, &slot, &offs)) {
-                if (spill) {
-                    /* Slot used in current instr may have a previous spilled gpr/aflags
-                     * value that may already have been restored. Forget that.
+            if (is_our_spill_or_restore(drcontext, inst, &spill, &reg, &slot, NULL)) {
+                if (!spill) {
+                    /* Found a restore for app aflags/gpr. While working our way back,
+                     * we'll look for the matching spill (identified using the slot id).
+                     * If we don't find any matching spill before the faulting instr, it
+                     * means that the spill happened before the fault occurred and needs
+                     * to be restored.
                      */
-                    spill_slot_to_reg[slot] = DR_REG_NULL;
-                    if (aflags_spill_reg == reg) {
-                        ASSERT(!gpr_native[GPR_IDX(aflags_spill_reg)],
-                               "reg with aflags cannot be native");
-                        spill_slot_to_reg[slot] = AFLAGS_ALIAS_REG;
-                    } else if (gpr_native[GPR_IDX(reg)]) {
-                        spill_slot_to_reg[slot] = reg;
-                    } else {
-                        LOG(drcontext, DR_LOG_ALL, 3,
-                            "%s @" PFX ": ignoring tool spill of non-native value.\n",
-                            __FUNCTION__, pc);
+                    if (reg == aflags_spill_reg) {
+                        ASSERT(spill_slot[GPR_IDX(AFLAGS_ALIAS_REG)] == MAX_SPILLS,
+                               "no spill found for last seen aflags restore");
+                        spill_slot[GPR_IDX(AFLAGS_ALIAS_REG)] = slot;
+                        aflags_spill_reg = DR_REG_NULL;
+                    } else if (reg == tool_aflags_spill_reg) {
+                        /* Found a restore instr that reads tool aflags from a slot into
+                         * a reg. This should not be confused with a restore for the reg
+                         * itself; we do not record any slot for tool_aflags_spill_reg.
+                         */
+                        tool_aflags_spill_reg = DR_REG_NULL;
+                    } else if (spill_slot[GPR_IDX(reg)] == MAX_SPILLS) {
```

Comment:
An aflags spill is two step: load aflags into gpr and then spill gpr to slot. Whereas for gpr it's just one step -- spill gpr to slot.

To determine whether we need to record a gpr restore, we check whether there's no slot recorded yet (`spill_slot[GPR_IDX(reg)] == MAX_SPILLS` on L2093).

For aflags, we check similarly -- whether there's no **reg or slot** recorded (see L2111 --`aflags_spill_reg == DR_REG_NULL && spill_slot[GPR_IDX(AFLAGS_ALIAS_REG)] == MAX_SPILLS`). When we get to the second part of aflags spill (the slot write), we expect there to be no slot recorded still.

Reply:
> For GPR you ignore restores you've already seen

For GPR we ignore restores if there's already a pending restore (there's some restore for which we haven't found a matching spill yet -- `spill_slot[GPR_IDX(reg)] != MAX_SPILLS`). We do the same for aflags below on L2111.

> for aflags you assert you never see another restore.

As explained in comment above, the assert on L2083 is because if `aflags_spill_reg` was recorded we expect `spill_slot[GPR_IDX(AFLAGS_ALIAS_REG)] == MAX_SPILLS` to be true. This is because it was a part of our precondition for recording aflags_spill_reg on L2111.